### ISSUE TITLE
Present/Past roots in lemma field

### ISSUE THREAD

**Issue by MemduhG** [UniversalDependencies/UD_Persian#1]
I've been looking through UD Persian as I'm working on two related languages, Kurmanji and Sorani. I've noticed that both past and present roots are shown in the lemma field separated by a pound sign `#` for a lot of verbs, like  _kardan_ below..
```
20   کرد     کرد#کن  VERB    V_PA    Number=Sing|Person=3|Tense=Past 0       root    _       _
```
(the lemma and root fields are swapped above due to RTL rendering)
Is this a temporary solution, or will it remain for UD Persian? And is this something that other languages with similar past/present roots for verbs should be doing as well?

**Comment by mojgan-seraji** [comment]
I have used the hash character to separate the past- and present lemmas for verbs in Persian UD. I never discussed it with anyone, as I didn't find it important. If this would remain as a solution for Persian UD, it should be applied to other languages with similar grammatical characteristics as well. Otherwise, another suggestion need to be presented.

**Comment by jnivre** [comment]
I don't know anything about past and present roots in Persian, but I don't think any other treebank has lemmas of this kind, so I think it would be better for cross-lingual consistency if we could find a solution where the lemma is always a single form. 

@dan-zeman Are you aware of anything similar in another language?

**Comment by dan-zeman** [comment]
Nope. Nothing I'm aware of. The assumption with lemmas is that they are the base/citation form of the word. Obviously there is no universal rule saying what the base form is, and language-particular tradition plays a role. I do not even think that a bare root shall be considered a lemma, unless the root is used in the language also as a full word form. (Some languages that do not have the infinitive use 1st person singular indicative present as the lemma.) But I believe that only one form should represent the lemma, not a combination of two forms.

I would suggest to use a present form, unless there are Persian-specific reasons to see the past forms as less marked.

**Comment by mojgan-seraji** [comment]
I personally do not call the present- and past lemmas as roots. Both share the same root and mean the same. In contemporary Persian, each verb has a present- and a past lemma. Although both lemmas share a single root, having access to one lemma does not predict the other. The present lemma is used for verbs in present and imperative forms, and the past lemma is used for verbs in past and future forms. Both lemmas are used as full word forms in the language. The present lemma is basically an imperative form (with omitted the letter 'b' at the beginning, as the word is an old word form remained from Old/Middle Persian, I believe) of 2ed person singular, and the past lemma is a past tense simple verb in 3rd person singular. In both cases, no personal ending is involved. 

Examples:
neveštan (to write) infinitive form in Contemporary Persian
nevisidan (to write) infinitive form in Old/Middle Persian - the infinitive form only exists in old scripts while the present form is frequently in use to build verbs in present and imperative forms.

nevis (write.​pres.​2sg)​​ = present lemma​
man mi-nevis-am (I write)
I cont-write.pres-1sg

neveš​t (write.past.3sg) = past lemma​​​
​man neveš​t-am (I wrote)
​I write.past​-1sg

**Comment by MemduhG** [comment]
I think in most related languages the lemma is thought of as the
past/infinitive, so I don't think choosing a present tense one would make a
lot of sense. We have been using the infinitives as lemmas in both Sorani
and Kurmanji.

However I do think that representing both past and present root for verbs
like this might be helpful for machine learning of morphology like in
UDPipe, as it would probably be easier to make the connection between
minevisam "I write" and nevis, the present stem as opposed to navasht,
the past one. One could argue that this happens in English/German etc as
well, but it's not exactly a matter of irregularity, the verbs tend to
follow certain patterns with their differrnces between past and present.

I do not know exactly how feasible representing this info would be for UD
though.

**Comment by ftyers** [comment]
I think the right thing here would be to have the infinitive _-an_ form as the lemma and then have the past/present stems in the Misc column. CoNLL-U format (if I understand correctly) discourages "technical" lemmas.

**Comment by dan-zeman** [comment]
@mojgan-seraji : I think this might be a terminological misunderstanding. I believe we understand lemma as one of the word forms of one lexeme, that is selected to represent the whole lexeme. Seen from this angle, there is no room for a verb to have more than one lemma. (Unless we say that the past forms are different lexeme, i.e. different verb than the present forms. Which I hope we do not want to say.)

Agreed with @MemduhG that there is a number of languages where some forms cannot be predicted from the lemma; but these languages still select just one form as the lemma. Slavic languages would also be members of that club, cf. Czech

lemma _psát_ (write):
_psá-t_ (to write, infinitive)
_psa-l_ (he wrote)
_píš-e_ (he writes)
_piš-0_ (write, imperative)

I agree that it can be useful to have the information about the alternating roots in the treebank but such information should go into the MISC column, not into LEMMA.

If Persian has infinitives, then I agree with @ftyers that the infinitive should be the lemma.

**Comment by jnivre** [comment]
I agree. For cross-linguistic consistency, the lemma should be a single form, preferably an infinitive. The lemma is primarily a representation of the lexeme and there is no requirement that other forms should be predictable from it.

**Comment by mojgan-seraji** [comment]
I understand our goal with UD but we cannot go for something incorrect to hide differences of language-specific constructions in the name of "cross-linguistic consistency".  

@dan-zeman : the reason of having two lemmas for each verb in Persian is because of the fact that each verb, in order to appear in different tenses relies on two different infinitives (as shown previously in my example), in other words, two different lexemes. The examples you, Dan, have shown in Czech for different tenses of the verb "write" come from one infinitive. This does not apply in Persian.

@jnivre : in the case of having an infinitive as the lemma, which one? There are two different infinitives and the lemma shifts depending on the tense of the verb.

**Comment by amir-zeldes** [comment]
I don't know much about Persian, but it seems to me that Classical Greek has the same situation: there are multiple stems which cannot be predicted from each other (present stem, perfect stem, future stem etc.) and they also have different infinitives (e.g. a perfect infinitive means 'to have done X').

The UD_Ancient_Greek data follows the convention in Classics, which is to use the present stem, although admittedly they use the first person singular present, rather than the infinitive. I also agree with @dan-zeman and @jnivre that a single lemma is preferable, with 'the other one' possibly being moved to misc as @ftyers suggested.

In general though, I'm not sure we should be forcing languages to give an infinitive as their lemma, since there are different traditions, and deviating from them will surprise and potentially discourage users. In Semitic languages, the convention is to use 3rd person singular masculine perfect as the lemma ('he wrote'), and that is in fact what UD_Arabic is doing for the lemma at the moment.

**Comment by ftyers** [comment]
I think rather than saying infinitive, it might be better to say citation (or dictionary) form. I wouldn't say that Persian (or Kurdish)[1] has two separate infinitives, rather that verb forms are formed from two separate stems, the infinitive is formed from the past stem + -tan/-dan.

In any case, Persian dictionaries are often (usually?) organised with the infinitive as the head word and then the present stem in brackets afterwards, for example this entry from Wiktionary:

```
کردن • ‎(kardan) ‎(present stem کن ‎(kon))
```


1. Armenian also has this  (pres) _kard-_  (past) _kardac῾-_ "to read", as does Ossetian (pres) _kən-_, (past) _kod-_ "do, make"

**Comment by rasoolims** [comment]
I also agree with @mojgan-seraji  about having both lemmas together. That's a pretty standard way to represent verbs in Persian. In creating another treebank for Persian ([Dadegan treebank](http://www.cs.columbia.edu/~rasooli/papers/Rasooli-et-al.,NAACL-HLT2013.pdf)), we discussed this issue with Persian linguistics and finally we realized that those lemmas can be easily distinguished with rules. As a byproduct I developed a rule-based verb analyzer that can give lemmas for verbs ([C# code](https://github.com/rasoolims/PersianVerbAnalyzer) and [Java code](https://github.com/rasoolims/PersianVerbAnalyzer-Java)). 

One more note: showing a verb with its infinitive is problematic and ambigious in some occasions:

رفتن (rftn) --> to go --> it should be رفت/رو (rft/rv)
رفتن (rftn) --> to sweep --> it should be رفت/روب  (rft/rvb)

**Comment by ftyers** [comment]
@rasoolims it can also be ambiguous in many other languages: English "lie" → "lay" , "lied". Is there a  widely used print dictionary that uses both stems[1] as a headword ? 

All of the other languages that I have looked at that have a similar past stem/present stem system to Persian (Armenian, Ossetian, Kurmanji, Tajik) use a single form based on the infinitive in their dictionaries.

---
1. I use "stem" here not lemma, as I consider _lemma_ to be "a single canonical form or citation form of a set of words". A _stem_ is a part of a word that affixes can be attached to. In some languages this may be the same as the lemma (a lot of the time in English: 'walk-' +s 'walks' +ed 'walked' etc.), but in other languages it may be different (Spanish 'habl-' +ar 'hablar' +o 'hablo' +as 'hablas' etc.).

**Comment by dan-zeman** [comment]
@mojgan-seraji : I certainly do not want you to do anything incorrect. We are just trying to unify terminology here. A lexeme is a unit of lexical meaning, an abstract entity covering all inflectional forms of a word. If _neveštan_ and _nevisidan_ belong to two different lexemes, then they are two different verbs (and morphology that relates them, if any, is just derivation, not inflection). One of the verbs would be used only in past clauses, the other only in present clauses. That would sound strange to me but I am no expert in Persian and I would accept it if you Persian linguists insist on it. However, then there would be no good reason for keeping both stems together in one LEMMA field. Present forms would have only _nevisid(an?)_ as their lemma, and past forms would have only _nevešt(an?)_ as their lemma. Because the lemma represents the lexeme, i.e. the set of forms that share one lemma. No word has two lemmas, by definition.

On the other hand, if the two sets of forms belong to just one lexeme (which seems more natural to me), it does not matter that there are two infinitive forms. FWIW, Finnish has about four distinct verb forms that are traditionally called infinitives. Czech has two forms of infinitives but they share the stem and they are semantically interchangeable; one of them is rather archaic. If you agree that _neveštan_ and _nevisidan_ are forms of one verb, then you can simply introduce a rule that says which of them will represent the lexeme as its lemma. My intuition is that a modern form should win over an Old/Middle Persian one—as long as UD Persian is a treebank of modern Persian :-)

**Comment by mojgan-seraji** [comment]
@dan-zeman: 
> My intuition is that a modern form should win over an Old/Middle Persian one—as long as UD Persian is a treebank of modern Persian :-)

Right on target! That was my intuition as well! 
Then, I go for the past lemma. This is also something that is close to what @amir-zeldes mentioned for the UD_Arabaic. 

Thank you all for the discussion.

**Comment by MemduhG** [comment]
I think the problem (or opportunity) where Persian, Kurdish and similar languages are concerned isn't about the irregular changes to the stem, but rather the existence of two (or more) different stems that can be quite regular, and indeed each can have the exact same person and number agreement morphemes as the others. This regularity might be something machine learned morph analyzers can take advantage of, if the information is represented. Some sort of standard of showing it in the misc column could make it easier to read for machines. Something like `Stem=kon` in the misc column comes to mind, but I am not sure how attractive the prospect of a feature like `Stem` with a potentially unlimited set of values would be.

**Comment by dan-zeman** [comment]
Unlimited values do not hurt in MISC. MISC features are not like FEAT features. You could think of MISC features as a means of introducing any number of additional "columns", all roofed by the MISC column.

**Comment by dan-zeman** [comment]
Fixed in 992e08a9978fccce83a768608bc00aef9a8b7d4e and surrounding commits (the fix was needed to be able to register the lemmas of auxiliaries with the validation infrastructure).
